app/beacon.py

```diff
@@ -61,156 +58,48 @@ def inner(*args, **kwargs):
-                total_balance += validator_balance
-                found_on_beacon_pubkeys += 1
+        df = pd.DataFrame.from_records([{**v, **v['validator']} for v in all_validators])
+        lido_validators = df.loc[df['pubkey'].isin(validator_pub_keys)]
 
-                if validator['status'] in ['active', 'active_ongoing']:
-                    active_validators_balance += validator_balance
+        validators_count = len(lido_validators)
+        total_balance = int(lido_validators['balance'].astype('int').sum())
 
-                # logging.info(f'Pubkey: {pubkey[:12]} Balance: {validator_balance} Gwei')  # todo uncomment
-            elif validator['status'] == 'UNKNOWN':
-                logging.warning(f'Pubkey {pubkey[:12]} status UNKNOWN')
+        active_validators = lido_validators.loc[lido_validators['status'].isin([ValidatorStatus.ACTIVE, ValidatorStatus.ACTIVE_ONGOING])]
+        active_validators_balance = int(active_validators['balance'].astype('int').sum())
 
         # Convert Gwei to wei
         total_balance *= 10 ** 9
         active_validators_balance *= 10 ** 9
 
-        return total_balance, found_on_beacon_pubkeys, active_validators_balance
-
+        return total_balance, validators_count, active_validators_balance
 
-class Prysm:
-    api_version = 'eth/v1alpha1/node/version'
-    api_genesis = 'eth/v1alpha1/node/genesis'
-    api_beacon_head = 'eth/v1alpha1/beacon/chainhead'
-    api_get_balances = 'eth/v1alpha1/validators/balances'
+    @staticmethod
+    def _from_bytes_to_pub_keys(keys_list):
+        return list(set('0x' + binascii.hexlify(key).decode() for key in keys_list))
 
-    def __init__(self, url, slots_per_epoch):
-        self.url = url
-        self.slots_per_epoch = slots_per_epoch
-        self.version = session.get(urljoin(url, self.api_version), timeout=DEFAULT_TIMEOUT).json()
-
-    @proxy_connect_timeout_exception
-    def get_finalized_epoch(self):
-        finalized_epoch = int(session.get(urljoin(self.url, self.api_beacon_head), timeout=DEFAULT_TIMEOUT).json()['finalizedEpoch'])
-        return finalized_epoch
-
-    @proxy_connect_timeout_exception
-    def get_genesis(self):
-        genesis_time = session.get(urljoin(self.url, self.api_genesis), timeout=DEFAULT_TIMEOUT).json()['genesisTime']
-        genesis_time = datetime.datetime.strptime(genesis_time, '%Y-%m-%dT%H:%M:%SZ').replace(tzinfo=timezone.utc)
-        genesis_time = int(genesis_time.timestamp())
-        return genesis_time
-
-    @proxy_connect_timeout_exception
-    def get_actual_slot(self):
-        actual_slots = {}
-        response = session.get(urljoin(self.url, self.api_beacon_head), timeout=DEFAULT_TIMEOUT).json()
-        actual_slots['actual_slot'] = int(response['headSlot'])
-        actual_slots['finalized_slot'] = int(response['finalizedSlot'])
-        return actual_slots
-
-    @proxy_connect_timeout_exception
-    def get_balances(self, slot, key_list):
-        params = {}
-        pubkeys = []
```

Comment:
Wasn't there a problem here? It seems that the last time we solved the performance problem, only the lighthouse version was fixed.

The problem is that searching in the array at each iteration in the loop is a very costly operation, since we are searching in an array of 160+ keys